tutorial.md

```diff
@@ -0,0 +1,305 @@
+    
+![toggle](http://localhost:8888/files/Screen%20Shot%202022-05-18%20at%203.43.49%20PM.png?_xsrf=2%7C7f8215bb%7C78c6f1df99528f5087b9d85f0eb5de75%7C1650662527)
+    
+Then, trigger it to run it:
+    
+![astro-simple.py](http://localhost:8888/files/Screen%20Shot%202022-05-18%20at%203.39.49%20PM.png?_xsrf=2%7C7f8215bb%7C78c6f1df99528f5087b9d85f0eb5de75%7C1650662527)
+    
+Click on the astro_orders DAG name to see the tree view of its execution:
+
+![tree-view](http://localhost:8888/files/Screen%20Shot%202022-05-18%20at%204.29.08%20PM.png?_xsrf=2%7C7f8215bb%7C78c6f1df99528f5087b9d85f0eb5de75%7C1650662527)
+    
+***
+# Walk through the code
+    
+## Extract
+    
+To extract from S3 into a Table object, we need only specify the location on S3 of the data and a connection for Snowflake that the Python SDK can use for internal purposes:
+    
+```python
+# Extract a file with a header from S3 into a Table object
+orders_data = aql.load_file(
+    # data file needs to have a header row
+    path=S3_FILE_PATH + '/orders_data_header.csv',
+    file_conn_id=S3_CONN_ID, output_table=Table(database=SNOWFLAKE_DATABASE,conn_id=SNOWFLAKE_CONN_ID)
+)   
+```
+## Transform
+    
+We can execute a filter and join in a single line of code to fill in a value for `joined_data`:
+    
+```python 
+@aql.transform
+def filter_orders (input_table: Table):
+   return "SELECT * FROM {{input_table}} WHERE amount > 150"
+
+@aql.transform
+def join_orders_customers (filtered_orders_table: Table, customers_table: Table):
+    return """SELECT c.customer_id, customer_name, order_id, purchase_date, amount, type 
+    FROM {{filtered_orders_table}} fo JOIN {{customers_table}} c 
+    ON fo.customer_id = c.customer_id"""
+
+# create a Table object for customer data in our Snowflake database
+customers_table =
+Table(table_name=SNOWFLAKE_CUSTOMERS,conn_id=SNOWFLAKE_CONN_ID, database=SNOWFLAKE_DATABASE)
+
+# filter the orders data and then join with the customer table
+joined_data = join_orders_customers(filter_orders(orders_data),customers_table)
+```
+    
+## Merge
+    
+As our penultimate step, we call a database-agnostic merge function:
+    
+```python
+    # merge the joined data into our reporting table, based on the order_id . 
+    # If there's a conflict in the customer_id or customer_name then use the ones from 
+    # the joined data
+    reporting_table = aql.merge(target_table =   Table(table_name=SNOWFLAKE_REPORTING,conn_id=SNOWFLAKE_CONN_ID,database=SNOWFLAKE_DATABASE,schema=SNOWFLAKE_SCHEMA,warehouse=SNOWFLAKE_WAREHOUSE),     
+        merge_table=joined_data,
+        merge_columns=["customer_id", "customer_name"],
+        target_columns=["customer_id", "customer_name"],
+        merge_keys={"order_id": "order_id"},
+        conflict_strategy="update")
+```   
+
+## Dataframe transformation
+
+As an illustration of the `@df` decorator, we show a simple dataframe operation:
+    
+```python
+@df
```

Comment:
What if we were consistent with the other decorators and call this `aql.dataframe`?

Reply:
+1. I was wondering why we have `@aql.transform` and `@aql.run_raw_sql` vs. `@df`

Reply:
cc @utkarsharma2 @dimberman -- Thoughts? Can one of you make this change tomorrow if you agree please?

Reply:
@kaxil @utkarsharma2 @dimberman if you're happy with it, I will move: `astro.dataframe` to `astro.sql.dataframe`, since it is used in the context of SQL.

We can then replace all references from:
```
from astro import dataframe as df
```
To:
```
from astro.sql import dataframe
```

Reply:
👍